Convert more peer peer REST calls

klauspost opened the pull request:
## Description

* Clean up in general.
* Add JSON wrapper.
* Add slice wrapper.
* Add option to make handler return nil error if no connection is given, `IgnoreNilConn`.

Converts the following:

```
+	HandlerGetMetrics
+	HandlerGetResourceMetrics
+	HandlerGetMemInfo
+	HandlerGetProcInfo
+	HandlerGetOSInfo
+	HandlerGetPartitions
+	HandlerGetNetInfo
+	HandlerGetCPUs
+	HandlerServerInfo
+	HandlerGetSysConfig
+	HandlerGetSysServices
+	HandlerGetSysErrors
+	HandlerGetAllBucketStats
+	HandlerGetBucketStats
+	HandlerGetSRMetrics
+	HandlerGetPeerMetrics
+	HandlerGetMetacacheListing
+	HandlerUpdateMetacacheListing
+	HandlerGetPeerBucketMetrics
+	HandlerStorageInfo
+	HandlerConsoleLog
+	HandlerGetLocks
+	HandlerBackgroundHealStatus
+	HandlerGetLastDayTierStats
+	HandlerSignalService
+	HandlerGetBandwidth
```

@harshavardhana We should probably postpone this until after next release. Even though I've tested there could be subtle things.

## How to test this PR?

Most are related to health diag, so `mc support diag` and `mc admin info` apply, but other calls apply as well.

`mc admin logs` appear broken on the current version. Needs https://github.com/minio/madmin-go/pull/269 to work.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

klauspost committed:
Convert more peer peer REST calls

* Clean up in general.
* Add JSON wrapper.
* Add slice wrapper.
* Add option to make handler return nil error if no connection is given, `IgnoreNilConn`.

Converts the following:

```
+	HandlerGetMetrics
+	HandlerGetResourceMetrics
+	HandlerGetMemInfo
+	HandlerGetProcInfo
+	HandlerGetOSInfo
+	HandlerGetPartitions
+	HandlerGetNetInfo
+	HandlerGetCPUs
+	HandlerServerInfo
+	HandlerGetSysConfig
+	HandlerGetSysServices
+	HandlerGetSysErrors
+	HandlerGetAllBucketStats
+	HandlerGetBucketStats
+	HandlerGetSRMetrics
+	HandlerGetPeerMetrics
+	HandlerGetMetacacheListing
+	HandlerUpdateMetacacheListing
+	HandlerGetPeerBucketMetrics
+	HandlerStorageInfo
```

klauspost committed:
Port ConsoleLog and more cleaning.

klauspost committed:
Document new types

klauspost committed:
Silly linter

klauspost committed:
Merge branch 'master' into convert-peer-rest-calls

harshavardhana committed:
Merge branch 'master' into convert-peer-rest-calls

harshavardhana committed:
Merge branch 'master' into convert-peer-rest-calls

klauspost committed:
Merge branch 'master' into convert-peer-rest-calls

klauspost committed:
Merge branch 'master' into convert-peer-rest-calls

# Conflicts:
#	cmd/peer-rest-client.go
#	cmd/peer-rest-common.go
#	cmd/peer-rest-server.go

klauspost committed:
Add extra disconnect safety

Fix reported races that are actually synchronized by network calls.

But this should add some extra safety for untimely disconnects.

Race reported:

```
WARNING: DATA RACE
Read at 0x00c00171c9c0 by goroutine 214:
  github.com/minio/minio/internal/grid.(*muxClient).addResponse()
      e:/gopath/src/github.com/minio/minio/internal/grid/muxclient.go:519 +0x111
  github.com/minio/minio/internal/grid.(*muxClient).error()
      e:/gopath/src/github.com/minio/minio/internal/grid/muxclient.go:470 +0x21d
  github.com/minio/minio/internal/grid.(*Connection).handleDisconnectClientMux()
      e:/gopath/src/github.com/minio/minio/internal/grid/connection.go:1391 +0x15b
  github.com/minio/minio/internal/grid.(*Connection).handleMsg()
      e:/gopath/src/github.com/minio/minio/internal/grid/connection.go:1190 +0x1ab
  github.com/minio/minio/internal/grid.(*Connection).handleMessages.func1()
      e:/gopath/src/github.com/minio/minio/internal/grid/connection.go:981 +0x610

Previous write at 0x00c00171c9c0 by goroutine 1081:
  github.com/minio/minio/internal/grid.(*muxClient).roundtrip()
      e:/gopath/src/github.com/minio/minio/internal/grid/muxclient.go:94 +0x324
  github.com/minio/minio/internal/grid.(*muxClient).traceRoundtrip()
      e:/gopath/src/github.com/minio/minio/internal/grid/trace.go:74 +0x10e4
  github.com/minio/minio/internal/grid.(*Subroute).Request()
      e:/gopath/src/github.com/minio/minio/internal/grid/connection.go:366 +0x230
  github.com/minio/minio/internal/grid.(*SingleHandler[go.shape.*github.com/minio/minio/cmd.DiskInfoOptions,go.shape.*github.com/minio/minio/cmd.DiskInfo]).Call()
      e:/gopath/src/github.com/minio/minio/internal/grid/handlers.go:554 +0x3fd
  github.com/minio/minio/cmd.(*storageRESTClient).DiskInfo()
      e:/gopath/src/github.com/minio/minio/cmd/storage-rest-client.go:314 +0x270
  github.com/minio/minio/cmd.erasureObjects.getOnlineDisksWithHealingAndInfo.func1()
      e:/gopath/src/github.com/minio/minio/cmd/erasure.go:293 +0x171
```

This read will always happen after the write, since there is a network call in between.

However a disconnect could come in while we are setting up the call, so we protect against that with extra checks.

klauspost committed:
Improve linter happiness

klauspost committed:
Merge branch 'convert-peer-rest-calls' of github.com:klauspost/minio into convert-peer-rest-calls

klauspost committed:
Convert ListDir storage RPC + clean up.

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -2004,49 +2003,70 @@ func (a adminAPIHandlers) ConsoleLogHandler(w http.ResponseWriter, r *http.Reque
 	// Avoid reusing tcp connection if read timeout is hit
 	// This is needed to make r.Context().Done() work as
-	// expected in case of read timeout
+	// expected in case of read timeout (????)
```
This logic is a bit dated, not sure if its still valid.

klauspost replied:
Yeah. Seems outdated.

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -2004,49 +2003,70 @@ func (a adminAPIHandlers) ConsoleLogHandler(w http.ResponseWriter, r *http.Reque
+	// expected in case of read timeout (????)
 	w.Header().Set("Connection", "close")
 
 	setEventStreamHeaders(w)
```
We only need to do this that's all no other HTTP headers needed.

harshavardhana commented on cmd/peer-rest-common.go:
```diff
@@ -18,68 +18,29 @@
 	peerRESTMethodCommitBinary                = "/commitbinary"
 	peerRESTMethodSignalService               = "/signalservice"
 	peerRESTMethodBackgroundHealStatus        = "/backgroundhealstatus"
 	peerRESTMethodGetLocks                    = "/getlocks"
```
These can be converted as well

harshavardhana commented on cmd/peer-rest-common.go:
```diff
@@ -18,68 +18,29 @@
-	peerRESTMethodLoadTransitionTierConfig    = "/loadtransitiontierconfig"
 	peerRESTMethodSpeedTest                   = "/speedtest"
 	peerRESTMethodDriveSpeedTest              = "/drivespeedtest"
 	peerRESTMethodReloadSiteReplicationConfig = "/reloadsitereplicationconfig"
```
and these as well, they do not send any data.

klauspost replied:
I was hesitant for the speedtests.. but I guess the information only ones would be fine.

klauspost replied:
Took the Reload handler.

harshavardhana commented on internal/grid/handlers.go:
```diff
@@ -409,6 +452,17 @@ func (h *SingleHandler[Req, Resp]) AllowCallRequestPool(b bool) *SingleHandler[R
+	if h.ignoreNilConn {
+		logger.LogOnceIf(context.Background(), fmt.Errorf("%s: IgnoreNilConn called twice", h.id.String()), h.id.String()+"IgnoreNilConn")
+	}
+	h.ignoreNilConn = true
```
what is the use-case for this?

klauspost replied:
This allows removing the `if conn == nil {return nil}` - hit when grid has not yet been initialized - in the client callers. See the changes in all the existing rpc calls.

klauspost committed:
Merge branch 'master' of https://github.com/minio/minio

klauspost committed:
Merge branch 'master' into convert-peer-rest-calls

# Conflicts:
#	cmd/peer-rest-client.go
#	cmd/peer-rest-common.go
#	cmd/peer-rest-server.go
#	internal/grid/handlers_string.go

klauspost committed:
Add

```
+	HandlerGetLocks
+	HandlerBackgroundHealStatus
+	HandlerGetLastDayTierStats
+	HandlerSignalService
+	HandlerGetBandwidth
```
